api/rtx/src/box_init.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "uvisor-lib/uvisor-lib.h"
+#include "mbed_interface.h"
+#include "cmsis_os.h"
+#include <stdint.h>
+#include <string.h>
+
+/* This function is called by uVisor in unprivileged mode. On this OS, we
+ * create box main threads for the box. */
+void __uvisor_lib_box_init(void * lib_config)
+{
+    osThreadId thread_id;
+    osThreadDef_t * flash_thread_def = lib_config;
+    osThreadDef_t * thread_def;
+
+    /* The thread definition is most likely in flash, so we need to copy it to
+     * box-local RAM before we can modify it. Allocate some RAM for storing the
+     * main thread's thread definition. This memory is never freed. */
+    thread_def = malloc_p(sizeof(*thread_def));
+    if (thread_def == NULL) {
+        /* No process heap memory available */
+        mbed_die();
+    }
+
+    /* Copy thread definition from flash to RAM. */
+    memcpy(thread_def, flash_thread_def, sizeof(*thread_def));
```

Comment:
Below the code you quoted, you can see that uVisor Lib modifies the provided `thread_def` before giving it to the RTOS (by allocating a stack from the tier-2 allocator). The RTOS uses the `thread_def` after our modifications. The RTX implementation of CMSIS RTOS uses the `thread_def` to call `rt_tsk_create`, which populates a separate in-ram TCB with various options.